packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,262 @@
+import * as React from 'react';
+import { unstable_batchedUpdates } from 'react-dom';
+import { HTMLElementWithStyledMap, getMotionDuration } from '../utils/dom-style';
+import { useAnimationFrame, useTimeout, usePrevious } from '@fluentui/react-utilities';
+
+export type MotionOptions = {
+  /**
+   * Whether to animate the element on first mount.
+   *
+   * @default false
+   */
+  animateOnFirstMount?: boolean;
+};
+
+export type MotionType = 'unmounted' | 'entering' | 'entered' | 'idle' | 'exiting' | 'exited';
+
+export type MotionState<Element extends HTMLElement = HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.Ref<Element>;
+
+  /**
+   * Current state of the element.
+   *
+   * - `unmounted` - The element is not yet rendered or can be safely removed from the DOM.
+   * - `entering` - The element is performing enter animation.
+   * - `entered` - The element has finished enter animation.
+   * - `idle` - The element is currently not animating, but rendered on screen.
+   * - `exiting` - The element is performing exit animation.
+   * - `exited` - The element has finished exit animation.
+   */
+  type: MotionType;
+
+  /**
+   * Indicates whether the component can be rendered.
+   * Useful to render the element before animating it or to remove it from the DOM after exit animation.
+   */
+  canRender: boolean;
```

Comment:
nit: I'd call this `shouldRender`

Reply:
What about the name being simply `render`? I feel that `shouldRender` looks like an "assert" kind of thing.